% Declarations for Front Matter

\title{Infrastructure for Scalable Analysis of Genomic Variation}
\author{Adam M. Novak}
\degreeyear{2017}
\degreemonth{June}
\degree{DOCTOR OF PHILOSOPHY}
\chair{Professor Josh Stuart}
\committeememberone{Distinguished Professor David Haussler}
\committeemembertwo{Assistant Professor Ed Green}
\committeememberthree{Associate Professor Beth Shapiro}
\numberofmembers{4} %% (including chair) possible: 3, 4, 5, 6
\deanlineone{Dean \textbf{Tyrus Miller}}
\deanlinetwo{Vice Provost and Dean of Graduate Studies}
\deanlinethree{}
\field{Bioinformatics}
\campus{Santa Cruz}

\begin{frontmatter}

\maketitle
\copyrightpage

\tableofcontents
\listoffigures
\listoftables

\begin{abstract}
The scale of the promlems which human genomics is asked to solve necessitates that the field develop an ability to integrate and synthesize information across the entire human population. The abstraction of a single-copy human reference genome assembly, and the linear coordinate space that it induces, are more of a hinderance than a help at these scales, because they can only ever represent one sample at any given place, and because they make combining information about human variation across multiple studies and modalities difficult. To rectify these problems, I propose the construction and adoption of a graph-based alternative to the human reference genome assembly: a Human Genome Variation Map. In the four research projects presented here, I present a theory of mapping to references that is extensible to graphs, describe a novel data structure for embedding individual haplotype sequences into a graph reference, survey graph construction techniques to discover methods that produce graphs yielding read mapping and variant calling results superior to those obtained with linear, variation-free references, and extend these improvement results to chromosome-scale graphs constructed from multiple sources and modalities of variation data. These four projects describe a research program aimed towards the eventual release of an official Human Genome Variation Map build, providing a piece of vital infrastructure for the analysis of human genomic variation at population scale.
\end{abstract}

\begin{acknowledgements}
In addition to my advisor, Prof. David Haussler, and the other members of my committee, I would like to thank Dr. Benedict Paten for helping to direct my research efforts, Erik Garrison for his leadership of the \vg team, Glenn Hickey for all the software plumbing he has written, Yohei Rosen for his amazing mathematics skills, Jordan Eizenga for his useful algorithms, Maciek Smuga-Otto and Sean Blum for their help on the GA4GH bakeoff project, Mike Lin for his technical challenges, and Microsoft Corporation for their provision of computing resources. I would also like to thank Anna Henderson for her contributions as an editor.

The text of this thesis includes reprints of the following previously published or preprint material:

{

\makeatletter
% We need this defined, with the standard definition from <https://tex.stackexchange.com/a/317947>, or latex doesn't know how to do the bibitems
\newcommand\newblock{\hskip .11em\@plus.33em\@minus.07em}
% We also need to basically redefine bibentry not to have a hyperlink target to not confuse links to the actual references.
% See <https://tex.stackexchange.com/a/141460>
\renewcommand\bibentry[1]{\nocite{#1}{\frenchspacing
     \@nameuse{BR@r@#1\@extra@b@citeb}}}
\makeatother

\begin{singlespace}
\noindent
\bibentry{novak2015canonical}
\end{singlespace}

In this work, I specifically helped develop the context scheme theory, wrote the majority of the custom software used in the analysis, produced the figures, and contributed substantially to the text of the manuscript.

\begin{singlespace}
\noindent
\bibentry{novak2016graph}
\end{singlespace}

In this work, I specifically developed the details of the eponymous graph extension and its algorithms, wrote most of the implementation code, ran the experiments, produced the figures, and contributed substantially to the text of the manuscript.

\begin{singlespace}
\noindent
\bibentry{novak2017genome}
\end{singlespace}

}

In this work, I specifically distributed the source data to participants, produced the Camel graph, developed and ran the reference-free evaluation, ran and analyzed the low-coverage read alignments, produced high-coverage alignments for variant calling, developed about half of the variant calling code, made numerous improvements and bug fixes to the \vg software to facillitate the analysis, and led the production of the manuscript.

The co-authors Benedict Paten and David Haussler listed in these publications directed and supervised the research which forms the basis for this thesis.
\end{acknowledgements}

\end{frontmatter}
